Emmeti infrared climate support

E440QF opened the pull request:
# What does this implement/fix?

This component adds support for Emmeti brand climate via infrared remote both in receiving and transmitting mode

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** not applicable

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3109

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
```yaml
climate:
  - platform: emmeti
    name: AC
    receiver_id: rcvr

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

E440QF committed:
Added preliminary support for Emmeti IR climate devices

E440QF committed:
Merge branch 'esphome:dev' into emmeti

E440QF committed:
set IR frequency

E440QF committed:
Fixed Emmeti IR transmission

E440QF committed:
Merge branch 'esphome:dev' into emmeti

E440QF committed:
emmeti changes

E440QF committed:
fixed support for receiving emmeti commands

E440QF committed:
using enums for better readability

E440QF committed:
added TODOs in comments

E440QF committed:
fix clang-format

E440QF committed:
updated CODEOWNERS file

E440QF committed:
added newline at the end of emmeti.h

E440QF committed:
fix clang-tidy

E440QF committed:
fix clang-tidy

E440QF committed:
added tests

jesserockz commented on esphome/components/emmeti/emmeti.cpp:
```diff
@@ -0,0 +1,368 @@
+  /*
+  if(!data.expect_item(EMMETI_BIT_MARK, EMMETI_MESSAGE_SPACE)){
+    return false;
+  }
+
+  for (size_t pos = 0; pos < 4; pos++){
+    if (data.expect_item(EMMETI_BIT_MARK, EMMETI_ONE_SPACE)){
+      curr_state.fan_pos |= 1 << pos;
+    }else if(!data.expect_item(EMMETI_BIT_MARK, EMMETI_ZERO_SPACE)){
+      return false;
+    }
+  }
+
+  for (size_t pos = 0; pos < 4; pos++){
+    if (!data.expect_item(EMMETI_BIT_MARK, EMMETI_ZERO_SPACE)){
+      return false;
+    }
+  }
+
+  for (size_t pos = 0; pos < 2; pos++){
+    if (data.expect_item(EMMETI_BIT_MARK, EMMETI_ONE_SPACE)){
+      curr_state.th |= 1 << pos;
+    }else if(!data.expect_item(EMMETI_BIT_MARK, EMMETI_ZERO_SPACE)){
+      return false;
+    }
+  }
+
+  for (size_t pos = 0; pos < 18; pos++){
+    if (!data.expect_item(EMMETI_BIT_MARK, EMMETI_ZERO_SPACE)){
+      return false;
+    }
+  }
+
+  for (size_t pos = 0; pos < 4; pos++){
+    if (data.expect_item(EMMETI_BIT_MARK, EMMETI_ONE_SPACE)){
+      curr_state.checksum |= 1 << pos;
+    }else if(!data.expect_item(EMMETI_BIT_MARK, EMMETI_ZERO_SPACE)){
+      return false;
+    }
+  }*/
```
No commented code please.
```suggestion
```

jesserockz commented on esphome/components/emmeti/emmeti.cpp:
```diff
@@ -0,0 +1,368 @@
+    this->mode = climate::CLIMATE_MODE_OFF;
+  }
+
+  /** @brief TODO: read checksum */
+  /*if(this->check_checksum_(curr_state.checksum)){
+    this->publish_state();
+    ESP_LOGD(TAG, "Published state");
+    return true;
+  }
+  return false;*/
```
Why is this commented?

E440QF replied:
The packet sent by the remote is in a format like
bits0-34 20ms_pause bits35-66
Because of the long pause, esphome doesn't recognize it as a single big packet, so i don't decode the second half (see [previously commented code](https://github.com/esphome/esphome/pull/5197#discussion_r1369242831)) this is not a big deal as that half of the packet doesn't hold any data we REALLY need.
The caveat is: the checksum is in the last 4 bits, so without receiving it we obviously cant verify it
```suggestion

```

E440QF committed:
remove currently dead commented code

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

E440QF committed:
remove dead code

E440QF committed:
Merge branch 'dev' into emmeti

E440QF committed:
fixed CODEOWNERS and clang-format

jesserockz commented on esphome/components/emmeti/emmeti.cpp:
```diff
@@ -0,0 +1,316 @@
+
+template<typename T>
+void EmmetiClimate::reverse_add_(T val, size_t len, esphome::remote_base::RemoteTransmitData *data) {
+  add_(reverse_(val, len), len, data);
```
Please ensure all class members and member functions are prefixed with `this->` 
```suggestion
  this->add_(reverse_(val, len), len, data);
```

jesserockz commented on esphome/components/emmeti/emmeti.cpp:
```diff
@@ -0,0 +1,316 @@
+bool EmmetiClimate::check_checksum_(uint8_t checksum) {
+  uint8_t expected = this->gen_checksum_();
+  ESP_LOGD(TAG, "Expected checksum: %X", expected);
+  ESP_LOGD(TAG, "Checksum received: %X", checksum);
```
```suggestion
  ESP_LOGV(TAG, "Expected checksum: %X", expected);
  ESP_LOGV(TAG, "Checksum received: %X", checksum);
```

jesserockz commented on esphome/components/emmeti/emmeti.h:
```diff
@@ -0,0 +1,109 @@
+const uint32_t EMMETI_ZERO_SPACE = 530;
+const uint32_t EMMETI_MESSAGE_SPACE = 20000;
+
+using state = struct State {
```
This is unnecessary
```suggestion
struct State {
```

E440QF committed:
Prepend this-> to all members in emmeti
Changed "using state = struct State" to "struct EmmetiState"
Use LOGV instead of LOGD for check_checksum_

jesserockz committed:
Merge branch 'dev' into pr/E440QF/5197

jesserockz committed:
Fix

jesserockz committed:
Move tests

jesserockz committed:
Fix mistake and add tx id
